### ISSUE TITLE
Add new interface OCP\Streamer

### ISSUE THREAD

**Issue by jakobsack** [nextcloud/server#7589]
To add the possibility to download all attachments of an email I'd like to use the following features from `OC\Streamer` and `OC_UTIL` in a public API:

```
OC_Util::obEnd();
$streamer = new Streamer();
$streamer->sendHeaders("attachments");
$streamer->addFileFromStream($fh, $attachment['fileName'], $attachment['size'], false);
$streamer->finalize();
```

The public interface could be called `OCP\Streamer`.
Origin of this issue is nextcloud/mail#580

**Comment by MorrisJobke** [comment]
cc @rullzer @icewind1991

**Comment by rullzer** [comment]
Fine by me. Would be good if that part got some more testing. And also would remove a private lib dependency of the files_sharing app.

**Comment by icewind1991** [comment]
That should probably be handled by the http response api

**Comment by rullzer** [comment]
@icewind1991 I like that idea!

**Comment by rullzer** [comment]
I can look into this for 14

**Comment by jakobsack** [comment]
Thank you @rullzer

**Comment by rullzer** [comment]
Mmmm so I have been thinking about this. If we go for the appframework responses (and I think we should). There are a few things to consider:

1. You can pass in resources. However they will only be read when constructing the output. So don't close the filehandle 😉 
2. I'd like to have the possibility as well to add the Node API `File` and `Folders` (just so we can use it as well).
3. We'd have to construct some temp FS.

So this would somehow lead to some functions:

```php

interface ZipResponse {

    public function addResource(resource $r, string $internalName, int $size, int $time = -1);

    public function addFile(File $f, string $internalName);

    public function addFolder(Folder $f, string $internalName);
}
```

This will just all be stored. When those functions are called. And we'll only construct the real zip once we want to create the output.

The zipstreamer by defaults outputs directly to `php://output` already. But we could chose to abstract away and have a mem/temp stream. But lets worry about that when we have to. I would have to check if the header stuff would not break down.

Anyways. I'll proceed with this soonish (tm)

**Comment by rullzer** [comment]
First POC in #8017

**Comment by MorrisJobke** [comment]
Looks like it will not make it into 14, right?

**Comment by rullzer** [comment]
My PoC is there but I guess not...

**Comment by rullzer** [comment]
Fixed with #10967
